packages/eslint-plugin/src/util/getFunctionHeadLoc.ts

```diff
@@ -0,0 +1,73 @@
+import {
+  AST_NODE_TYPES,
+  AST_TOKEN_TYPES,
+  TSESLint,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+type FunctionNode =
+  | TSESTree.ArrowFunctionExpression
+  | TSESTree.FunctionExpression
+  | TSESTree.FunctionDeclaration;
+
+/**
+ * Creates a report location for the given function.
+ * The location only encompasses the "start" of the function, and not the body
+ *
+ * eg.
+ *
+ * ```
+ * function foo(args) {}
+ * ^^^^^^^^^^^^^^^^^^
+ *
+ * get y(args) {}
+ * ^^^^^^^^^^^
+ *
+ * const x = (args) => {}
+ *           ^^^^^^^^^
+ * ```
+ */
+export function getFunctionHeadLoc(
```

Comment:
while we're here - one case I believe this code doesn't correctly handle is decorators because decorators are included within the method's start.

Eg this code will highlight:

```ts
class Foo {
  @deco
  ^^^^^
  foo() {}
  ^^^^^
}
```

this has confused some people in the past (eg https://github.com/typescript-eslint/typescript-eslint/issues/751)